### PR TITLE
Implement redis lock

### PR DESCRIPTION
**Pull request by nover**
This PR implements the Distributed Redis Application Lock.

### CODE REVIEW

File: src/rapidcore.redis.csproj [hunk 1]
```diff
@@ -1,22 +1,30 @@
...
-    <Company>SKARP ApS</Company>
-    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
-    <CodeAnalysisRuleSet>../stylecop.ruleset</CodeAnalysisRuleSet>
-  </PropertyGroup>
-
-</Project>
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>netstandard1.6</TargetFramework>
+    <RootNamespace>RapidCore</RootNamespace>
```
Comment (pingvinen) [inline src/rapidcore.redis.csproj]: Shouldn't RootNamespace be `RapidCore.Redis`?

File: src/rapidcore.redis.csproj [hunk 2]
```diff
@@ -1,22 +1,30 @@
...
+    <Version>0.0.0</Version>
+    <Title>RapidCore</Title>
```
Comment (pingvinen) [inline src/rapidcore.redis.csproj]: RapidCore.Redis?

File: test/functional/Locking/Redis/RedisDistributedAppLockerTest.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            // Create task to dispose of loclk at some point
+            Task.Factory.StartNew(() =>
+            {
+                Task.Delay(TimeSpan.FromMilliseconds(700));
+                // release the first lock
+                firstLock.Dispose();
+
+                // wait and the new lock should be acquired
+                Task.Delay(TimeSpan.FromMilliseconds(500));
+                Assert.Equal(lockName, secondLock.Name);
```
Comment (pingvinen) [inline test/functional/Locking/Redis/RedisDistributedAppLockerTest.cs]: These asserts look out of place, but I am a bit confused about this test. Is this the expected flow?
1. acquire firstLock
2. try to acquire second lock, but be patient (20 secs)
3. thread unlocks after 700 ms
4. thread waits 500 ms
5. now we expect the second call to acquire to have succeeded, so inside the thread we now assert stuff

If so, maybe we should add that as a comment to the test.

File: test/functional/functionaltest.csproj [hunk 4]
```diff
@@ -1,17 +1,15 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
-
   <PropertyGroup>
     <TargetFramework>netcoreapp1.1</TargetFramework>
+    <AssemblyName>functionaltest</AssemblyName>
+    <RootNamespace>RapidCore.FunctionalTest</RootNamespace>
```
Comment (pingvinen) [inline test/functional/functionaltest.csproj]: RapidCore.Redis.FunctionalTest

File: test/unit/Locking/Redis/RedisDistributedAppLockTest.cs [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+                A<TimeSpan>.Ignored,
+                A<CommandFlags>.Ignored)).Returns(true);
+
+            var handle = new RedisDistributedAppLock(manager);
+            handle.AcquireLockAsync(lockName, TimeSpan.FromSeconds(2));
+
+            A.CallTo(() => client.LockTake(
+                    A<RedisKey>.That.Matches(str => str == lockName),
+                    A<RedisValue>.Ignored,
+                    A<TimeSpan>.Ignored,
```
Comment (pingvinen) [inline test/unit/Locking/Redis/RedisDistributedAppLockTest.cs]: Isn't this test supposed to show that we pass on the timespan?

File: test/unit/Locking/Redis/RedisDistributedAppLockTest.cs [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+                A<CommandFlags>.Ignored)).Returns(true);
+
+            var handle = new RedisDistributedAppLock(manager);
+
+            using (await handle.AcquireLockAsync(lockName, TimeSpan.FromSeconds(2)))
+            {
+                // empty on purpose
+            }
+
+            // after using scope both client and underlying lock must be released
```
Comment (pingvinen) [inline test/unit/Locking/Redis/RedisDistributedAppLockTest.cs]: The client should be left intact right?

File: test/unit/Locking/Redis/RedisDistributedAppLockTest.cs [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+                 A<RedisKey>.That.Matches(str => str == lockName),
+                 A<RedisValue>.Ignored,
+                 A<TimeSpan>.Ignored,
+                 A<CommandFlags>.Ignored)).Throws(new Exception("test is faking it!"));
+
+            var handle = new RedisDistributedAppLock(manager);
+
+            var ex = await Assert.ThrowsAsync<DistributedAppLockException>(
+                async () => await handle.AcquireLockAsync(lockName, TimeSpan.FromSeconds(2)));
+            Assert.Equal(DistributedAppLockExceptionReason.SeeInnerException, ex.Reason);
```
Comment (pingvinen) [inline test/unit/Locking/Redis/RedisDistributedAppLockTest.cs]: Add assertion on the inner exception, so we prove that it is actually set.

File: src/Locking/Redis/RedisDistributedAppLock.cs [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+        /// <returns><c>this</c> upon successful lock grab (for a fluent interface)</returns>
+        /// <exception cref="DistributedAppLockException"></exception>
+        public async Task<IDistributedAppLock> AcquireLockAsync(string lockName, TimeSpan? timeout = null)
+        {
+            var stopWatch = new Stopwatch();
+            try
+            {
+                LockHandle = Guid.NewGuid().ToString("N");
+                var noTimeoutProvided = timeout == null;
+                if (noTimeoutProvided)
```
Comment (pingvinen) [inline src/Locking/Redis/RedisDistributedAppLock.cs]: Shouldn't this just be `timeout.HasValue`?

**Review by pingvinen — CHANGES_REQUESTED** [review]
The unit tests could use some constructors to make them more DRY. They seem to repeat the same setup in most of the facts.

File: src/rapidcore.redis.csproj (hunk 1 above)
Comment (nover) [reply src/rapidcore.redis.csproj]: My thought was that our locking primitives are in `RapidCore.Locking` so the Redis locks should be in `RapidCore.Locking.Redis` and not `RapidCore.Redis.Locking`...

File: test/functional/Locking/Redis/RedisDistributedAppLockerTest.cs (hunk 3 above)
Comment (nover) [reply test/functional/Locking/Redis/RedisDistributedAppLockerTest.cs]: This is exactly what happens

File: test/unit/Locking/Redis/RedisDistributedAppLockTest.cs (hunk 6 above)
Comment (nover) [reply test/unit/Locking/Redis/RedisDistributedAppLockTest.cs]: Yes the RedisMuxer should not be touched..

File: src/rapidcore.redis.csproj (hunk 1 above)
Comment (pingvinen) [reply src/rapidcore.redis.csproj]: I see your point, however that makes it slightly less obvious that it comes from the RapidCore.Redis package.

Comment (nover) [reply src/rapidcore.redis.csproj]: Decision time then: Root namespace of RapidCore assemblies should match the assembly name.

File: test/unit/Locking/Redis/RedisDistributedAppLockTest.cs (hunk 5 above)
Comment (nover) [reply test/unit/Locking/Redis/RedisDistributedAppLockTest.cs]: That timespan is the lock auto expiry - test is invalid after moving to StackExchange.Redis

**Review by pingvinen — APPROVED** [review]

**Review by MartinsRui — APPROVED** [review]

**Pull request merged**